<a href="https://colab.research.google.com/github/Saurav-1399/Coursera_Capstone/blob/main/Battle_of_the_Neighbourhoods_Toronto.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Battle of the Neighbourhoods - Toronto

Author: Saurav Menon M

This notebook contains Questions 1, 2 & 3 of the Assignment. They have been segregated by Section headers

In [ ]:
import pandas as pd

# Question 1

## Importing Data

In [ ]:
import requests

In [ ]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means that we are able to make the connection to the page

In [ ]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 5           M6A        North York   
 6           M7A  Downtown Toronto   
 7           M8A      Not assigned   
 8           M9A         Etobicoke   
 9           M1B       Scarborough   
 10          M2B      Not assigned   
 11          M3B        North York   
 12          M4B         East York   
 13          M5B  Downtown Toronto   
 14          M6B        North York   
 15          M7B      Not assigned   
 16          M8B      Not assigned   
 17          M9B         Etobicoke   
 18          M1C       Scarborough   
 19          M2C      Not assigned   
 20          M3C        North York   
 21          M4C         East York   
 22          M5C  Downtown Toronto   
 23          M6C              York   
 24          M7C      Not assigned   
 25         

In [ ]:
len(wiki_data), type(wiki_data)

(3, list)

We need the first table alone, so dropping the other tables

In [ ]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
..          ...               ...   
150         M7W      Not assigned   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
154         M2X      Not assigned   
155         M3X      Not assigned   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
158         M6X      Not assigned   
159         M7X      Not assigned   
160         M8X         Etobicoke   
161         M9X      Not assigned   
162         M1Y      Not assigned   
163         M2Y      Not assigned   
164         M3Y      Not assigned   
165         M4Y  Downtown Toronto   
166         M5Y      Not assigned   
167         M6Y      Not assigned   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
170         M9Y      Not assigned   
171         M1Z      Not assigned   
172         M2Z      Not assigned   
173         M3Z      Not assigned   
174         M4Z      Not assigned   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
7                                         Not assigned  
8              Islington Avenue, Humber Valley Village  
9                                       Malvern, Rouge  
10                                        Not assigned  
11                                           Don Mills  
12                     Parkview Hill, Woodbine Gardens  
13                            Garden District, Ryerson  
14                                           Glencairn  
15                                        Not assigned  
16                                        Not assigned  
17   West Deane Park, Princess Gardens, Martin Grov...  
18              Rouge Hill, Port Union, Highland Creek  
19                                        Not assigned  
20                                           Don Mills  
21                                    Woodbine Heights  
22                                      St. James Town  
23                                  Humewood-Cedarvale  
24                                        Not assigned  
25                                        Not assigned  
26   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
27                   Guildwood, Morningside, West Hill  
28                                        Not assigned  

Dropping Borough which are not assigned

In [ ]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
8           M9A         Etobicoke   
9           M1B       Scarborough   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
17          M9B         Etobicoke   
18          M1C       Scarborough   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
26          M9C         Etobicoke   
27          M1E       Scarborough   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
36          M1G       Scarborough   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
..          ...               ...   
111         M4R   Central Toronto   
112         M5R   Central Toronto   
113         M6R      West Toronto   
114         M7R       Mississauga   
116         M9R         Etobicoke   
117         M1S       Scarborough   
120         M4S   Central Toronto   
121         M5S  Downtown Toronto   
122         M6S      West Toronto   
126         M1T       Scarborough   
129         M4T   Central Toronto   
130         M5T  Downtown Toronto   
135         M1V       Scarborough   
138         M4V   Central Toronto   
139         M5V  Downtown Toronto   
142         M8V         Etobicoke   
143         M9V         Etobicoke   
144         M1W       Scarborough   
147         M4W  Downtown Toronto   
148         M5W  Downtown Toronto   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
8              Islington Avenue, Humber Valley Village  
9                                       Malvern, Rouge  
11                                           Don Mills  
12                     Parkview Hill, Woodbine Gardens  
13                            Garden District, Ryerson  
14                                           Glencairn  
17   West Deane Park, Princess Gardens, Martin Grov...  
18              Rouge Hill, Port Union, Highland Creek  
20                                           Don Mills  
21                                    Woodbine Heights  
22                                      St. James Town  
23                                  Humewood-Cedarvale  
26   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
27                   Guildwood, Morningside, West Hill  
30                                         The Beaches  
31                                         Berczy Park  
32                                 Caledonia-Fairbanks  
36                                              Woburn  
39                                             Leaside  
40                                  Central Bay Street  
41                                            Christie  
45                                           Cedarbrae  
46                                   Hillcrest Village  
47     Bathurst Manor, Wilson Heights, Downsview North  

Grouping the records based on Postal Code

In [ ]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
8           M9A         Etobicoke   
9           M1B       Scarborough   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
17          M9B         Etobicoke   
18          M1C       Scarborough   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
26          M9C         Etobicoke   
27          M1E       Scarborough   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
36          M1G       Scarborough   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
..          ...               ...   
111         M4R   Central Toronto   
112         M5R   Central Toronto   
113         M6R      West Toronto   
114         M7R       Mississauga   
116         M9R         Etobicoke   
117         M1S       Scarborough   
120         M4S   Central Toronto   
121         M5S  Downtown Toronto   
122         M6S      West Toronto   
126         M1T       Scarborough   
129         M4T   Central Toronto   
130         M5T  Downtown Toronto   
135         M1V       Scarborough   
138         M4V   Central Toronto   
139         M5V  Downtown Toronto   
142         M8V         Etobicoke   
143         M9V         Etobicoke   
144         M1W       Scarborough   
147         M4W  Downtown Toronto   
148         M5W  Downtown Toronto   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
8              Islington Avenue, Humber Valley Village  
9                                       Malvern, Rouge  
11                                           Don Mills  
12                     Parkview Hill, Woodbine Gardens  
13                            Garden District, Ryerson  
14                                           Glencairn  
17   West Deane Park, Princess Gardens, Martin Grov...  
18              Rouge Hill, Port Union, Highland Creek  
20                                           Don Mills  
21                                    Woodbine Heights  
22                                      St. James Town  
23                                  Humewood-Cedarvale  
26   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
27                   Guildwood, Morningside, West Hill  
30                                         The Beaches  
31                                         Berczy Park  
32                                 Caledonia-Fairbanks  
36                                              Woburn  
39                                             Leaside  
40                                  Central Bay Street  
41                                            Christie  
45                                           Cedarbrae  
46                                   Hillcrest Village  
47     Bathurst Manor, Wilson Heights, Downsview North  

Checking for number of records where Neighbourhood is "Not assigned"

In [ ]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [ ]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
5        8         M9A         Etobicoke   
6        9         M1B       Scarborough   
7       11         M3B        North York   
8       12         M4B         East York   
9       13         M5B  Downtown Toronto   
10      14         M6B        North York   
11      17         M9B         Etobicoke   
12      18         M1C       Scarborough   
13      20         M3C        North York   
14      21         M4C         East York   
15      22         M5C  Downtown Toronto   
16      23         M6C              York   
17      26         M9C         Etobicoke   
18      27         M1E       Scarborough   
19      30         M4E      East Toronto   
20      31         M5E  Downtown Toronto   
21      32         M6E              York   
22      36         M1G       Scarborough   
23      39         M4G         East York   
24      40         M5G  Downtown Toronto   
25      41         M6G  Downtown Toronto   
26      45         M1H       Scarborough   
27      46         M2H        North York   
28      47         M3H        North York   
29      48         M4H         East York   
..     ...         ...               ...   
73     111         M4R   Central Toronto   
74     112         M5R   Central Toronto   
75     113         M6R      West Toronto   
76     114         M7R       Mississauga   
77     116         M9R         Etobicoke   
78     117         M1S       Scarborough   
79     120         M4S   Central Toronto   
80     121         M5S  Downtown Toronto   
81     122         M6S      West Toronto   
82     126         M1T       Scarborough   
83     129         M4T   Central Toronto   
84     130         M5T  Downtown Toronto   
85     135         M1V       Scarborough   
86     138         M4V   Central Toronto   
87     139         M5V  Downtown Toronto   
88     142         M8V         Etobicoke   
89     143         M9V         Etobicoke   
90     144         M1W       Scarborough   
91     147         M4W  Downtown Toronto   
92     148         M5W  Downtown Toronto   
93     151         M8W         Etobicoke   
94     152         M9W         Etobicoke   
95     153         M1X       Scarborough   
96     156         M4X  Downtown Toronto   
97     157         M5X  Downtown Toronto   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
5              Islington Avenue, Humber Valley Village  
6                                       Malvern, Rouge  
7                                            Don Mills  
8                      Parkview Hill, Woodbine Gardens  
9                             Garden District, Ryerson  
10                                           Glencairn  
11   West Deane Park, Princess Gardens, Martin Grov...  
12              Rouge Hill, Port Union, Highland Creek  
13                                           Don Mills  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
18                   Guildwood, Morningside, West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                    

In [ ]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
..          ...               ...   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
5              Islington Avenue, Humber Valley Village  
6                                       Malvern, Rouge  
7                                            Don Mills  
8                      Parkview Hill, Woodbine Gardens  
9                             Garden District, Ryerson  
10                                           Glencairn  
11   West Deane Park, Princess Gardens, Martin Grov...  
12              Rouge Hill, Port Union, Highland Creek  
13                                           Don Mills  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
18                   Guildwood, Morningside, West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  Central Bay Street  
25                                            Christie  
26                                           Cedarbrae  
27                                   Hillcrest Village  
28     Bathurst Manor, Wilson Heights, Downsview North  

In [ ]:
df.shape

(103, 3)

#### Answer to Question 1: We have 103 rows and 3 columns

# Question 2

## Installing geocoder

In [ ]:
!pip install geocoder

     |████████████████████████████████| 102kB 8.9MB/s eta 0:00:01


In [ ]:
import geocoder

Tried the below approach, ran for 20 mins, then killed it. Changing the code cell to Text for now so that the run all execution doesn't stop.

##### initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

##### loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

Alternatively, as suggested in the assignment, Importing the CSV file from the URL

In [ ]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
..          ...        ...        ...
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [ ]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [ ]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [ ]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [ ]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
..          ...               ...   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5              Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                       Malvern, Rouge  43.806686 -79.194353  
7                                            Don Mills  43.745906 -79.352188  
8                      Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                             Garden District, Ryerson  43.657162 -79.378937  
10                                           Glencairn  43.709577 -79.445073  
11   West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724  
12              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
13                                           Don Mills  43.725900 -79.340923  
14                                    Woodbine Heights  43.695344 -79.318389  
15                                      St. James Town  43.651494 -79.375418  
16                                  Humewood-Cedarvale  43.693781 -79.428191  
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201  
18                   Guildwood, Morningside, West Hill  43.763573 -79.188711  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy 

In [ ]:
combined_data.shape

(103, 5)

#### Solution: We get 103 rows as expected when we do a inner join, so we have good data.

# Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [ ]:
from geopy.geocoders import Nominatim

In [ ]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [ ]:
!pip install folium

     |████████████████████████████████| 102kB 9.1MB/s ta 0:00:011


In [ ]:
import folium

In [ ]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [ ]:
CLIENT_ID = 'XDRQZ5F5BIHTJFQRHDVCWAFAGGZX3BTRIPVBRVLRP143YBFM' 
CLIENT_SECRET = 'R543DKQOABFRXBQD3FICDARFMKDYLGP4GAXDADDIUDNH01Z1'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: XDRQZ5F5BIHTJFQRHDVCWAFAGGZX3BTRIPVBRVLRP143YBFM
CLIENT_SECRET:R543DKQOABFRXBQD3FICDARFMKDYLGP4GAXDADDIUDNH01Z1


In [ ]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [ ]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [ ]:
venues_in_toronto.shape

(1320, 5)

So we have 1320 records and 5 columns. Checking sample data

In [ ]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

Checking the Venues based on Neighbourhood

In [ ]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
5                                      Victoria Village   
6                                      Victoria Village   
7                             Regent Park, Harbourfront   
8                             Regent Park, Harbourfront   
9                             Regent Park, Harbourfront   
10                            Regent Park, Harbourfront   
11                            Regent Park, Harbourfront   
37                     Lawrence Manor, Lawrence Heights   
38                     Lawrence Manor, Lawrence Heights   
39                     Lawrence Manor, Lawrence Heights   
40                     Lawrence Manor, Lawrence Heights   
41                     Lawrence Manor, Lawrence Heights   
50          Queen's Park, Ontario Provincial Government   
51          Queen's Park, Ontario Provincial Government   
52          Queen's Park, Ontario Provincial Government   
53          Queen's Park, Ontario Provincial Government   
54          Queen's Park, Ontario Provincial Government   
80                                       Malvern, Rouge   
81                                            Don Mills   
82                                            Don Mills   
83                                            Don Mills   
84                                            Don Mills   
85                      Parkview Hill, Woodbine Gardens   
86                      Parkview Hill, Woodbine Gardens   
87                      Parkview Hill, Woodbine Gardens   
...                                                 ...   
1193                Northwest, West Humber - Clairville   
1194                        St. James Town, Cabbagetown   
1195                        St. James Town, Cabbagetown   
1196                        St. James Town, Cabbagetown   
1197                        St. James Town, Cabbagetown   
1198                        St. James Town, Cabbagetown   
1224             First Canadian Place, Underground city   
1225             First Canadian Place, Underground city   
1226             First Canadian Place, Underground city   
1227             First Canadian Place, Underground city   
1228             First Canadian Place, Underground city   
1254      The Kingsway, Montgomery Road, Old Mill North   
1255      The Kingsway, Montgomery Road, Old Mill North   
1256                               Church and Wellesley   
1257                               Church and Wellesley   
1258                               Church and Wellesley   
1259                               Church and Wellesley   
1260                               Church and Wellesley   
1286  Business reply mail Processing Centre, South C...   
1287  Business reply mail Processing Centre, South C...   
1288  Business reply mail Processing Centre, South C...   
1289  Business reply mail Processing Centre, South C...   
1290  Business reply mail Processing Centre, South C...   
1302  Old Mill South, King's Mill Park, Sunnylea, Hu...   
1303  Old Mill South, King's Mill Park, Sunnylea, Hu...   
1304  Mimico NW, The Queensway West, South of Bloor,...   
1305  Mimico NW, The Queensway West, South of Bloor,...   
1306  Mimico NW, The Queensway West, South of Bloor,...   
1307  Mimico NW, The Queensway West, South of Bloor,...   
1308  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude  \
0                  43.753259               -79.329656   
1                  43.753259               -79.329656   
2                  43.725882               -79.315572   
3                  43.725882               -79.315572   
4                  43.725882               -79.315572   
5                  43.725882             

So there are 413 records for each neighbourhood.

Checking for the maximum venue categories

In [ ]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                                     
Accessories Store                                              Wexford, Maryvale   
Airport                                                                Downsview   
Airport Food Court             CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate                   CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge                 CN Tower, King and Spadina, Railway Lands, Har...   
Airport Service                CN Tower, King and Spadina, Railway Lands, Har...   
Airport Terminal               CN Tower, King and Spadina, Railway Lands, Har...   
American Restaurant                     Toronto Dominion Centre, Design Exchange   
Antique Shop                                       High Park, The Junction South   
Aquarium                       Harbourfront East, Union Station, Toronto Islands   
Art Gallery                             Toronto Dominion Centre, Design Exchange   
Art Museum                                                    Central Bay Street   
Arts & Crafts Store                                  Willowdale, Willowdale East   
Asian Restaurant                        Toronto Dominion Centre, Design Exchange   
Athletics & Sports                                              Woodbine Heights   
Auto Garage                                                    Wexford, Maryvale   
Auto Workshop                  Business reply mail Processing Centre, South C...   
BBQ Joint                                    The Annex, North Midtown, Yorkville   
Baby Store                                                              Christie   
Bagel Shop                     Summerhill West, Rathnelly, South Hill, Forest...   
Bakery                                                         Wexford, Maryvale   
Bank                                              University of Toronto, Harbord   
Bar                                               University of Toronto, Harbord   
Baseball Field                 Old Mill South, King's Mill Park, Sunnylea, Hu...   
Basketball Stadium             Harbourfront East, Union Station, Toronto Islands   
Beer Bar                                          University of Toronto, Harbord   
Beer Store                                                      Woodbine Heights   
Belgian Restaurant                     Kensington Market, Chinatown, Grange Park   
Bike Shop                                                                Leaside   
Bistro                                                               Berczy Park   
...                                                                          ...   
Sporting Goods Shop                            North Toronto West, Lawrence Park   
Sports Bar                                                               Leaside   
Stadium                             Brockton, Parkdale Village, Exhibition Place   
Stationery Store                                                 Studio District   
Steakhouse                                           Willowdale, Willowdale East   
Supermarket                                                     Thorncliffe Park   
Supplement Shop                Mimico NW, The Queensway West, South of Bloor,...   
Sushi Restaurant                                     Willowdale, Willowdale East   
Swim School                                                        Lawrence Park   
Tailor Shop                                                       Stn A PO Boxes   
Taiwanese Restaurant                                 St. James Town, Cabbagetown   
Tanning Salon                  Mimico NW, The Queensway West, South of Bloor,...   
Tea Room                                  First Canadian Place, Underground city   
Tennis Court                                                  Humewood-Cedarvale   
Thai Restaurant                                                  Studio District   
Theater                            

There are around 237 different types of Venue Categories. Interesting!

# One Hot encoding the venue Categories

In [ ]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
5                     0        0                   0             0   
6                     0        0                   0             0   
7                     0        0                   0             0   
8                     0        0                   0             0   
9                     0        0                   0             0   
10                    0        0                   0             0   
11                    0        0                   0             0   
12                    0        0                   0             0   
13                    0        0                   0             0   
14                    0        0                   0             0   
15                    0        0                   0             0   
16                    0        0                   0             0   
17                    0        0                   0             0   
18                    0        0                   0             0   
19                    0        0                   0             0   
20                    0        0                   0             0   
21                    0        0                   0             0   
22                    0        0                   0             0   
23                    0        0                   0             0   
24                    0        0                   0             0   
25                    0        0                   0             0   
26                    0        0                   0             0   
27                    0        0                   0             0   
28                    0        0                   0             0   
29                    0        0                   0             0   
...                 ...      ...                 ...           ...   
1290                  0        0                   0             0   
1291                  0        0                   0             0   
1292                  0        0                   0             0   
1293                  0        0                   0             0   
1294                  0        0                   0             0   
1295                  0        0                   0             0   
1296                  0        0                   0             0   
1297                  0        0                   0             0   
1298                  0        0                   0             0   
1299                  0        0                   0             0   
1300                  0        0                   0             0   
1301                  0        0                   0             0   
1302                  0        0                   0             0   
1303                  0        0                   0             0   
1304                  0        0                   0             0   
1305                  0        0                   0             0   
1306                  0        0                   0             0   
1307                  0        0                   0             0   
1308                  0        0                   0             0   
1309                  0        0                   0             0   
1310                  0        0                   0             0   
1311                  0        0                   0             0   
1312                  0        0                   0             0   
1313                  0        0                   0             0   
1314                  0        0                   0             0   
1315            

Adding the neighbourhood to the encoded dataframe

In [ ]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  Turkish Restaurant  \
0                    0             0  ...              0                   0   
1                    0             0  ...              0                   0   
2                    0             0  ...              0                   0   
3                    0             0  ...              0                   0   
4                    0             0  ...              0                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 238 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [ ]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0             0.000000           0.0  ...            0.0   
1               0.0             0.000000           0.0  ...            0.0   
2               0.0             0.000000           0.0  ...            0.0   
3               0.0             0.000000           0.0  ...            0.0   
4               0.0             0.045455           0.0  ...            0.0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 238 columns]

Let's make a function to get the top most common venue categories

In [ ]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [ ]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [ ]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Bank   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1             Sandwich Place              Pharmacy           Coffee Shop   
2                       Bank              Pharmacy    Chinese Restaurant   
3         Chinese Restaurant   Japanese Restaurant                  Café   
4             Sandwich Place    Italian Restaurant       Thai Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store    Dim Sum Restaurant           Escape Room   
1                   Pub                   Gym       Airport Service   
2         Shopping Mall        Sandwich Place                 Diner   
3           Yoga Studio    Dim Sum Restaurant           Escape Room   
4              Pharmacy           Pizza Place                   Pub   

  8th Most Common Venue        9th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant   
1  Ethiopian Restaurant            Electronics Store   
2            Restaurant                Deli / Bodega   
3     Electronics Store  Eastern European Restaurant   
4                  Café                   Restaurant   

        10th Most Common Venue  
0          Dumpling Restaurant  
1  Eastern European Restaurant  
2                  Supermarket  
3          Dumpling Restaurant  
4                      Butcher

Let's make the model to cluster our Neighbourhoods

In [ ]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [ ]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
    random_state=0, tol=0.0001, verbose=0)

Checking the labelling of our model

In [ ]:
kmeans.labels_[0:100]

array([3, 1, 1, 1, 1, 1, 1, 1, 3, 1, 4, 1, 3, 1, 1, 1, 1, 3, 1, 1, 3, 3,
       1, 1, 3, 3, 4, 1, 1, 1, 3, 1, 3, 3, 3, 1, 3, 3, 1, 2, 3, 1, 1, 1,
       3, 1, 3, 1, 1, 3, 3, 3, 3, 1, 3, 1, 3, 1, 2, 1, 1, 4, 1, 1, 1, 4,
       3, 2, 1, 3, 3, 1, 1, 1, 1, 3, 1, 1, 1, 3, 1, 3, 1, 1, 1, 1, 0, 1,
       4, 3, 1, 1, 1, 3, 4, 3], dtype=int32)

Let's add the clustering Label column to the top 10 common venue categories

In [ ]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [ ]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             4.0                  Park   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio           Dessert Shop   
1           Coffee Shop          Hockey Arena  Portuguese Restaurant   
2                  Park                Bakery         Breakfast Spot   
3     Accessories Store              Boutique              Gift Shop   
4           Yoga Studio                 Diner         Sandwich Place   

    5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0             Escape Room     Electronics Store  Eastern European Restaurant   
1            Intersection             Drugstore                   Donut Shop   
2                     Pub        Chocolate Shop          Distribution Center   
3  Furniture / Home Store           Event Space                  Coffee Shop   
4   Portuguese Restaurant                  Park           Mexican Restaurant   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0   Dumpling Restaurant              Drugstore             Donut Shop  
1   Dumpling Restaurant          Deli / Bodega    Distribution Center  
2                   Spa         Farmers Market           Dessert Shop  
3         Women's Store  Vietnamese Restaurant       Airport Terminal  
4    Italian Restaurant             Hobby Shop    Fried Chicken Joint

Drop all the NaN values to prevent data skew

In [ ]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [ ]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [ ]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [ ]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             0.0            Print Shop           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11      Department Store           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

   9th Most Common Venue 10th Most Common Venue  
11            Donut Shop                Dog Run

Cluster 2

In [ ]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1         North York             1.0           Pizza Place   
2   Downtown Toronto             1.0           Coffee Shop   
3         North York             1.0        Clothing Store   
4   Downtown Toronto             1.0           Coffee Shop   
7         North York             1.0                   Gym   
8          East York             1.0           Pizza Place   
9   Downtown Toronto             1.0                  Café   
13        North York             1.0                   Gym   
15  Downtown Toronto             1.0             Gastropub   
17         Etobicoke             1.0           Pizza Place   
20  Downtown Toronto             1.0           Coffee Shop   
22       Scarborough             1.0           Coffee Shop   
23         East York             1.0           Coffee Shop   
24  Downtown Toronto             1.0           Coffee Shop   
25  Downtown Toronto             1.0         Grocery Store   
28        North York             1.0           Coffee Shop   
29         East York             1.0        Sandwich Place   
30  Downtown Toronto             1.0                  Café   
33        North York             1.0        Clothing Store   
34        North York             1.0    Miscellaneous Shop   
36  Downtown Toronto             1.0                 Plaza   
37      West Toronto             1.0                   Bar   
38       Scarborough             1.0            Hobby Shop   
39        North York             1.0                  Bank   
41      East Toronto             1.0      Greek Restaurant   
42  Downtown Toronto             1.0           Coffee Shop   
43      West Toronto             1.0                  Café   
47      East Toronto             1.0                  Park   
48  Downtown Toronto             1.0                  Café   
50        North York             1.0           Pizza Place   
54      East Toronto             1.0           Coffee Shop   
55        North York             1.0           Coffee Shop   
58       Scarborough             1.0       College Stadium   
59        North York             1.0      Ramen Restaurant   
65       Scarborough             1.0     Indian Restaurant   
70         Etobicoke             1.0           Pizza Place   
72        North York             1.0           Pizza Place   
73   Central Toronto             1.0           Coffee Shop   
74   Central Toronto             1.0                  Café   
75      West Toronto             1.0        Breakfast Spot   
76       Mississauga             1.0                 Hotel   
79   Central Toronto             1.0          Dessert Shop   
80  Downtown Toronto             1.0                  Café   
81      West Toronto             1.0                  Café   
82       Scarborough             1.0           Pizza Place   
84  Downtown Toronto             1.0                  Café   
86   Central Toronto             1.0           Coffee Shop   
87  Downtown Toronto             1.0        Airport Lounge   
88         Etobicoke             1.0           Coffee Shop   
89         Etobicoke             1.0           Pizza Place   
90       Scarborough             1.0  Fast Food Restaurant   
93         Etobicoke             1.0           Pizza Place   
96  Downtown Toronto             1.0    Italian Restaurant   
97  Downtown Toronto             1.0                  Café   
99  Downtown Toronto             1.0               Gay Bar   

     2nd Most Common Venue   3rd Most Common Venue    4th Most Common Venue  \
1              Coffee Shop            Hockey Arena    Portuguese Restaurant   
2                     Park                  Bakery           Breakfast Spot   
3        Accessories Store                Boutique                Gift Shop   
4              Yoga Studio                   Diner           Sandwich Place   
7               Beer Store             Coffee Shop      Japanese Restaurant   
8                Gastropub                Pharmacy                

Cluster 3

In [ ]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
12   Scarborough             2.0  Construction & Landscaping   
57    North York             2.0              Baseball Field   
101    Etobicoke             2.0  Construction & Landscaping   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12                    Bar           Yoga Studio    Dim Sum Restaurant   
57            Yoga Studio    Falafel Restaurant  Ethiopian Restaurant   
101        Baseball Field           Yoga Studio    Dim Sum Restaurant   

    5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
12   Ethiopian Restaurant           Escape Room            Electronics Store   
57            Escape Room     Electronics Store  Eastern European Restaurant   
101  Ethiopian Restaurant           Escape Room            Electronics Store   

           8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
12   Eastern European Restaurant   Dumpling Restaurant              Drugstore  
57           Dumpling Restaurant             Drugstore             Donut Shop  
101  Eastern European Restaurant   Dumpling Restaurant              Drugstore

Cluster 4

In [ ]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
6         Scarborough             3.0  Fast Food Restaurant   
10         North York             3.0           Pizza Place   
14          East York             3.0          Skating Rink   
16               York             3.0          Tennis Court   
18        Scarborough             3.0     Electronics Store   
19       East Toronto             3.0          Neighborhood   
26        Scarborough             3.0  Caribbean Restaurant   
27         North York             3.0    Athletics & Sports   
31       West Toronto             3.0                Bakery   
32        Scarborough             3.0            Playground   
40         North York             3.0         Grocery Store   
44        Scarborough             3.0              Bus Line   
45         North York             3.0   Martial Arts School   
46         North York             3.0         Grocery Store   
49         North York             3.0                  Park   
51        Scarborough             3.0   American Restaurant   
53         North York             3.0         Grocery Store   
56               York             3.0            Restaurant   
60         North York             3.0         Grocery Store   
61    Central Toronto             3.0                  Park   
62    Central Toronto             3.0           Music Venue   
63               York             3.0        Breakfast Spot   
67    Central Toronto             3.0               Dog Run   
68    Central Toronto             3.0         Jewelry Store   
69       West Toronto             3.0                   Bar   
71        Scarborough             3.0     Accessories Store   
77          Etobicoke             3.0           Pizza Place   
78        Scarborough             3.0                Lounge   
83    Central Toronto             3.0            Playground   
85        Scarborough             3.0            Playground   
92   Downtown Toronto             3.0              Beer Bar   
94          Etobicoke             3.0             Drugstore   
98          Etobicoke             3.0                  Pool   
100      East Toronto             3.0    Light Rail Station   
102         Etobicoke             3.0                   Gym   

          2nd Most Common Venue 3rd Most Common Venue  \
6                   Yoga Studio           Event Space   
10                         Park                   Pub   
14                 Intersection           Curling Ice   
16                        Field          Hockey Arena   
18                         Bank            Restaurant   
19                          Pub     Health Food Store   
26             Hakka Restaurant    Athletics & Sports   
27                  Golf Course                  Pool   
31                     Pharmacy           Music Venue   
32                   Smoke Shop         Jewelry Store   
40                         Park          Liquor Store   
44                       Bakery                  Park   
45                  Yoga Studio          Dessert Shop   
46                         Park          Liquor Store   
49   Construction & Landscaping                Bakery   
51                        Motel           Yoga Studio   
53                         Park          Liquor Store   
56           Turkish Restaurant                   Bar   
60                         Park          Liquor Store   
61                  Swim School              Bus Line   
62                       Garden           Yoga Studio   
63            Convenience Store               Brewery   
67            Food & Drink Shop          Dance Studio   
68                         Park      Sushi Restaurant   
69           Mexican Restaurant       Thai Restaurant   
71    Middle Eastern Restaurant            Smoke Shop   
77                         Park              Bus Line   
78    Latin American Restaurant        Breakfast Spot   
83                        Trail           Yoga Studio   
85                         Park    

Cluster 5

In [ ]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             4.0                  Park   
21              York             4.0                  Park   
35         East York             4.0                  Park   
64              York             4.0                  Park   
66        North York             4.0                  Park   
91  Downtown Toronto             4.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop           Yoga Studio          Dessert Shop   
21         Women's Store                  Pool           Yoga Studio   
35     Convenience Store          Intersection           Yoga Studio   
64           Yoga Studio      Department Store           Escape Room   
66     Convenience Store           Yoga Studio          Dessert Shop   
91                 Trail            Playground           Yoga Studio   

   5th Most Common Venue        6th Most Common Venue  \
0            Escape Room            Electronics Store   
21      Department Store            Electronics Store   
35          Dessert Shop                  Escape Room   
64     Electronics Store  Eastern European Restaurant   
66           Escape Room            Electronics Store   
91          Dessert Shop           Dim Sum Restaurant   

          7th Most Common Venue        8th Most Common Venue  \
0   Eastern European Restaurant          Dumpling Restaurant   
21  Eastern European Restaurant          Dumpling Restaurant   
35            Electronics Store  Eastern European Restaurant   
64          Dumpling Restaurant                    Drugstore   
66  Eastern European Restaurant          Dumpling Restaurant   
91                        Diner               Discount Store   

   9th Most Common Venue 10th Most Common Venue  
0              Drugstore             Donut Shop  
21             Drugstore             Donut Shop  
35   Dumpling Restaurant              Drugstore  
64            Donut Shop                Dog Run  
66             Drugstore             Donut Shop  
91   Distribution Center             Donut Shop

#### We have successfully cluster Toronto neighbourhood based on venue categories!